File: crates/taflow-core/src/stream/lag.rs
//! Causal delay-line state.

use std::collections::VecDeque;

use super::{operator_states::validate_period, StreamingIndicator};
use crate::TaResult;

/// Delays a scalar series by a fixed number of bars.
#[derive(Debug, Clone)]
/// Persistent Rust state or aligned output type for `Lag`.
///
/// The state consumes chronological inputs causally, preserves warm-up
/// values, and exposes the current result through its public API.
pub struct Lag {
    values: VecDeque<f64>,
    timeperiod: usize,
    value: Option<f64>,
}

impl Lag {
    /// Creates an empty delay line.
    ///
    /// `timeperiod` is the number of bars before the first value is emitted.
    pub fn new(timeperiod: usize) -> TaResult<Self> {
        validate_period(timeperiod)?;
        Ok(Self {
            values: VecDeque::with_capacity(timeperiod),
            timeperiod,
            value: None,
        })
    }

    /// Appends one observation and returns the value from `timeperiod` bars ago.
    pub fn append(&mut self, input: f64) -> Option<f64> {
        self.value = if self.values.len() == self.timeperiod {
            let value = self.values.pop_front().expect("lag window is full");
            self.values.push_back(input);
            Some(value)
        } else {
            self.values.push_back(input);
            None
        };
        self.value
    }

    /// Append a slice into `output` with `NaN` at warm-up positions.
    pub fn extend_slice_into(&mut self, input: &[f64], output: &mut Vec<f64>) {
        output.reserve(input.len());
        output.extend(
            input
                .iter()
                .map(|&input| self.append(input).unwrap_or(f64::NAN)),
        );
    }

    /// Returns the latest delayed value, or `None` during warm-up.
    pub fn value(&self) -> Option<f64> {
        self.value
    }

    /// Clears buffered observations and the latest value.
    pub fn reset(&mut self) {
        self.values.clear();
        self.value = None;
    }
}

impl StreamingIndicator for Lag {
    type Output = f64;

    fn append(&mut self, input: f64) -> Option<Self::Output> {
        Self::append(self, input)
    }

    fn value(&self) -> Option<Self::Output> {
        Self::value(self)
    }

    fn reset(&mut self) {
        Self::reset(self);
    }
}

File: python/taflow/lag.py
"""Persistent causal lag indicator."""

from typing import Any

import numpy as np

from ._native import Lag as _NativeLag
from ._series import as_float64_series


class Lag:
    """Return the value from ``timeperiod`` bars earlier using Rust state.

    ``_input`` is required; pass an empty series for a fresh streaming state.
    ``timeperiod`` defaults to 1 and must be positive. The first
    ``timeperiod`` history positions are ``NaN``. Correctness maps to pandas
    ``Series.shift``.
    """

    def __init__(self, _input: Any, timeperiod: int = 1) -> None:
        self._state = _NativeLag(timeperiod)
        self.extend(_input)

    def append(self, _input: float) -> "Lag":
        """Append one observation and return this indicator."""
        self._state.append(float(_input))
        return self

    def extend(self, _input: Any) -> "Lag":
        """Append chronological observations and return this indicator."""
        self._state.extend(as_float64_series(_input))
        return self

    def compute(self) -> np.ndarray:
        """Return the aligned history, including ``NaN`` warm-up."""
        return self._state.compute()

    @property
    def value(self) -> float | None:
        """Return the latest delayed value, or ``None`` during warm-up."""
        return self._state.value

    def reset(self) -> "Lag":
        """Restore fresh native state and return this indicator."""
        self._state.reset()
        return self

    def __len__(self) -> int:
        """Return the number of processed observations."""
        return len(self._state)
